README.md
# Keypoint Detection
Our model uses a convolutional neural network with a recurrent module to detect body keypoints from a single image. Below, we provide code to train and test the model.

For a web-based demo visit the [project page](http://www.robots.ox.ac.uk/~vgg/software/keypoint_detection/).

### Software Dependencies
You need MATLAB and the [MatConvNet](https://github.com/vlfeat/matconvnet) toolbox to run this demo. This demo has been compiled and tested for Matlab 2016a, CUDNN v5.1 and cuda 8.0 (using a Linux machine).

### Train Model (MPII Human Pose Dataset)
1. Download and prepare the dataset by executing **getMPIIData_v3.m** and then **splitMPIIData_V4.m**.
	- The data preparation takes several minutes to be completed. The output is the train (**MPI_imdbsT1aug0.mat**, ~3GB) and validation (**MPI_imdbsV1aug0.mat**,  ~0.3GB) files.   
2. Execute **trainBodyPose_example.m** (need to set some parameters such as MatConvNet path and GPU index).
3. To test model, run the demo that follows is explained below.

Parameters for training a model with one recurrent iteration:
```
net = initialize3ObjeRecFusion(opts,2,0,'shareFlag',[0,1]);

opts.derOutputs = {'objective1', 1,'objective2', 1, 'objective4', 1,'objective5', 1, 'objective7', 1,'objective8', 1};

```

Parameters for training a model with two recurrent iterations (**default**):
```
net = initialize3ObjeRecFusion(opts,3,0,'shareFlag',[0,1,1]);

opts.derOutputs = {'objective1', 1,'objective2', 1, 'objective4', 1,'objective5', 1, 'objective7', 1,'objective8', 1, 'objective10', 1,'objective11', 1};

```

### Run Demo
 1. Download a [pre-trained model](https://github.com/ox-vgg/keypoint_models/tree/master/models) if you haven't trained one.
 2. Execute **demo_keypoint.m**. 

### Run Live Demo
You need to have the **Web-Camera** package installed and set the path to it.

 1. Download a [pre-trained model](https://github.com/ox-vgg/keypoint_models/tree/master/models) if you haven't trained one.
 2. Execute **demolive_keypoints.m**. 

## Citation

    @inproceedings{Belagiannis17,
            title={Recurrent Human Pose Estimation},
            author={Belagiannis, Vasileios and Zisserman, Andrew},
            booktitle={International Conference on Automatic Face and Gesture Recognition},
            year={2017},
            organization={IEEE}
    }


For further details on the model and web-based demo, please visit our [project-page](http://www.robots.ox.ac.uk/~vgg/software/keypoint_detection/).

## License

Copyright (c) 2017, University of Oxford
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


CHANGELOG
keypoint-v1 : Initial model file (epoch 30) [July 14, 2016] 
keypoint-v2 : Recurret model as described in https://arxiv.org/abs/1605.02914 [Aug. 18, 2016]
keypoint-v3 : Recurret model with 2 iterations as described in http://www.robots.ox.ac.uk/~vgg/publications/2017/Belagiannis17/ [March. 23, 2017]
keypoint-v4 : Recurret model with 2 iterations (trained on MSCOCO, then fined-tuned in MPII Pose as described in http://www.robots.ox.ac.uk/~vgg/publications/2017/Belagiannis17/ [March. 23, 2017]

All the old model files can be downloaded from https://gitlab.com/vggdemo/keypoint_data

trainBodyPose_example.m
function trainBodyPose_example(varargin)
%% Keypoints detection: Training on MPII Human Pose Dataset
% This is a MatConvNet demo for human pose human estimation.
% Related Work: Belagiannis V., and Zisserman A.,
% Recurrent Human Pose Estimation, FG (2017).
% Contact: Vasileios Belagiannis, vb@robots.ox.ac.uk

%% Read Me
% Before executing the training script, the MPII dataset hat to be
% downloaded using the following scripts:
% 1. Download and prepare MPII Pose Dataset by executing:
% getMPIIData_v3.m and then splitMPIIData_V4.m (may take several minutes)
% To train a model the following has to be defined:
% - MatConvNet directories
% - Number of GPU (opts.gpus)

%% MatConvNet library setup (execute vl_setupnn.m)
run(fullfile(fileparts(mfilename('fullpath')),...
  '..', 'matconvnet-b23','matlab', 'vl_setupnn.m')) ;

%% Pose Example
 
addpath('dagnetworks');
addpath('model-train');

opts.datas='MPI';

%actual input to the network (after augmentation)
opts.patchHei=248;
opts.patchWi=248;

opts.flipFlg='mpi';%model for flipping the joint (flip augmentation)
opts.cam=1;
opts.aug=0;
opts.NoAug=1; %used for calling the correct imdb creation function

opts.batchSize = 20;
opts.numEpochs = 50 ;
opts.learningRate = [0.00001*ones(1, 30) 0.000005*ones(1, 5) 0.000001*ones(1, 50)] ;
opts.batchNormalization = 1;%useful for big networks

%GPU
opts.gpus = [3];

opts.outNode=16;%if heatmaps loss, then number of heatmaps
opts.outPairNode=15;% pairwise terms
opts.inNode=3;
opts.lossFunc='l2loss-heatmap';
opts.lossFunc2='l2loss-pairwiseheatmap';
opts.lossFunc3=[];
opts.ConcFeat=384;  %number of channels at concat
opts.skip_layer = 'layer20'; %skip layer

%export path, imdb store path and location of training / validation data.
opts.expDir = sprintf('../data/v1.00-%s_%s_%d_2Obje3Fus',opts.datas,opts.lossFunc,opts.cam) ;
opts.imdbPath = sprintf('../data/%s-baseline_imdb%d.mat',opts.datas, opts.cam);
opts.DataMatTrain=sprintf('../data/%s_imdbsT%daug%d.mat',opts.datas,opts.cam,opts.aug);
opts.DataMatVal=sprintf('../data/%s_imdbsV%daug%d.mat',opts.datas,opts.cam,opts.aug);

%transformation from input image (248X248) to ouput heatmap(62X62)
trf=[0.25 0 0 ; 0 0.25 0; 0 0 1]; %only scale

%build network - 1 iteration
%net = initialize3ObjeRecFusion(opts,2,0,'shareFlag', [0,1]);

%objectives
%opts.derOutputs = {'objective1', 1,'objective2', 1, ... %feed-forward
%             'objective4', 1,'objective5', 1, ... %iter 0 (not-shared w)
%             'objective7', 1,'objective8', 1};    %iter 1 (shared w)
 


%build network - 2 iterations
net = initialize3ObjeRecFusion(opts,3,0,'shareFlag',[0,1,1]);

%objectives
opts.derOutputs = {'objective1', 1,'objective2', 1, ... %feed-forward
             'objective4', 1,'objective5', 1, ... %iter 0 (not-shared w)
             'objective7', 1,'objective8', 1, ... %iter 1 (shared w)
             'objective10', 1,'objective11', 1};  %iter 2 (shared w)

         
opts.net=net;

opts.numThreads = 15;
opts.transformation = 'f25' ;
opts.averageImage = single(repmat(128,1,1,opts.inNode));
opts.fast = 1;
opts.imageSize = [248, 248] ;
opts.border = [8, 8] ;
opts.bord=[0,0,0,0]; %cropping border

%heatmap setting;
opts.heatmap=1;
opts.trf=trf;
opts.sigma=1.3;
opts.FiltSize=31;
opts.HeatMapSize=[62, 62];
opts.padGTim=[0 0];
opts.rotate=1;%rotation flag
opts.scale=1;%scale augm.
%extra parse settings

%occluded keypoints
opts.inOcclud=1;

%multiple instances
opts.multipInst=0;

%heatmap scheme
opts.HeatMapScheme=1; %how to generate heatmaps

opts.train.momentum=0.95;

opts.negHeat=0;%set to 1 to include negative values for the occlusion
opts.ignoreOcc=0;%set to 1 to include negative values for the occlusion
opts.ignoreRest=0; %quasi single human training

opts.pairHeatmap=1; %generate heatmaps for pairs of body parts
opts.bodyPairs = [1 2 3 4 5 7 8 9 11 12 13 14 14 15 7; 2 3 7 5 6 4 10 10 12 13 8 8 15 16 8]; %full body - MPI

opts.magnif=12;%amplifier for the body heatmaps
opts.facX=0.15;%pairwise heatmap width (def. 0.15)
opts.facY=0.08;%pairwise heatmap height

%imdb generation function
opts.imdbfn =@getImdbNoAug;

opts = vl_argparse(opts, varargin);

%create imdb and train
cnn_regressor_dag(opts);
